core/drivers/atmel_trng.c

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Microchip
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <libfdt.h>
+#include <matrix.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <string.h>
+#include <tee/tee_cryp_utl.h>
+
+/* Registers */
+#define TRNG_CTRL		0x0
+#define TRNG_CTRL_WAKEY_OFFSET	8
+#define TRNG_CTRL_WAKEY_VALUE	0x524E47
+
+#define TRNG_IER		0x10
+#define TRNG_ISR		0x1C
+#define TRNG_ODATA		0x50
+
+static vaddr_t trng_base;
+static uint32_t trng_last_val;
+static uint8_t trng_valid_bytes;
+
+static uint32_t atmel_trng_read32(void)
+{
+	while (!io_read32(trng_base + TRNG_ISR))
+		;
+
+	return io_read32(trng_base + TRNG_ODATA);
+}
+
+TEE_Result crypto_rng_read(void *buf, size_t len)
+{
+	uint8_t *rngbuf = buf;
+	uint32_t val = 0;
+	uint8_t len_to_copy = 0;
+
+	assert(buf);
+	assert(trng_base);
+
+	while (len) {
+		val = atmel_trng_read32();
```

Comment:
Do we need to protect this function from concurrent calls?

Reply:
ISR is a cleared on read register so even if two thread enters this function, only one of them will read 0 and then read ODATA. The second one will read ISR after that and will also get a data from odata. ISR will only be set after reading ODATA. So there is need to protect this function.

Reply:
I guess you meant to say that there isn't a need to protect this function.
What if the CPU breaking out of the loop is interrupted before it's able to read ODATA and then resumes at the same time as the other CPU is breaking out of the loop?
It seems more safe to rely on synchronization mechanisms designed for the purpose.

Reply:
Sorry, I meant that there were no need to protect this indeed.

> What if the CPU breaking out of the loop is interrupted before it's able to read ODATA and then resumes at the same time as the other CPU is breaking out of the loop?

If this happens,  the other CPU will not drop out of the loop since ISR won't be set until first CPU reads ODATA. As I said, ISR is set only **after** reading ODATA and ISR is cleared on read.

I can add a spinlock if needed but according to the hardware specification, there is normally no need for it.

Reply:
Is the hardware specification publicly available?

Reply:
Yes: https://ww1.microchip.com/downloads/en/DeviceDoc/DS60001476B.pdf
Section 63.

Reply:
Hum actually, after reading it again, you might be right. It seems like the ISR will be trigged again even if not reading ODATA.

Reply:
> 

Thanks